FHIR/src/Microsoft.Health.Fhir.Anonymizer.Shared.Core/Visitors/AnonymizationVisitor.cs

```diff
@@ -85,34 +89,128 @@ private ProcessResult ProcessResourceNode(ElementNode node)
+
+                LogProcessResult(node, rule, resultOnRule);
+
+                result.Update(resultOnRule);
+            }
+
+            return result;
+        }
+
+        private void InitializeNodeCache(ITypedElement node)
+        {
+            _typeToNodeCache.Clear();
+            _nameToNodeCache.Clear();
+
+            InitializeNodeCacheRecursive(node);
+        }
+
+        private void InitializeNodeCacheRecursive(ITypedElement node)
+        {
+            foreach (var child in node.Children())
+            {
+                // Cache instance type
+                if (_typeToNodeCache.ContainsKey(child.InstanceType))
                 {
-                    /** 
-                     * Special case handling:
-                     * Senario: FHIR path only contains resourceType: Patient, Resource. 
-                     * Sample AnonymizationFhirPathRule: { "path": "Patient", "method": "keep" }
-                     * 
-                     * Current FHIR path lib do not support navigate such ResourceType FHIR path from resource in bundle.
-                     * Example: navigate with FHIR path "Patient" from "Bundle.entry[0].resource[0]" is not support
-                     */
-                    matchNodes = new List<ElementNode>() { node };
+                    _typeToNodeCache[child.InstanceType].Add(child);
                 }
                 else
                 {
-                    matchNodes = node.Select(rule.Expression).CastElementNodes();
+                    _typeToNodeCache[child.InstanceType] = new List<ITypedElement> { child };
                 }
-                
-                foreach (var matchNode in matchNodes)
+
+                // Cache name
+                if (_nameToNodeCache.ContainsKey(child.Name))
                 {
-                    resultOnRule.Update(ProcessNodeRecursive(matchNode, _processors[method], context, rule.RuleSettings));
+                    _nameToNodeCache[child.Name].Add(child);
+                }
+                else
+                {
+                    _nameToNodeCache[child.Name] = new List<ITypedElement> { child };
                 }
-                LogProcessResult(node, rule, resultOnRule);
 
-                result.Update(resultOnRule);
+                // Recursively process node's children except resource children, which will be processed independently as a resource
+                if (!child.IsFhirResource())
+                {
+                    InitializeNodeCacheRecursive(child);
+                }
             }
+        }
 
-            return result;
+        private IEnumerable<ITypedElement> GetMatchNodes(AnonymizationFhirPathRule rule, ITypedElement node)
+        {
+            var matchNodes = new List<ITypedElement>();
+            var resourceTypeMatch = AnonymizationFhirPathRule.TypeRuleRegex.Match(rule.Path);
+            var nameMatch = AnonymizationFhirPathRule.NameRuleRegex.Match(rule.Path);
+
+            if (resourceTypeMatch.Success)
```

Comment:
Can we reuse the logic in matching by type and name here?

Reply:
Sure. A private function is added to hold the shared logics.